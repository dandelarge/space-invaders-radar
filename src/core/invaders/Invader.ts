import { PrintInvader } from "../PrintInvader";
import { RadarResults } from "../radar.types";

export class Invader {
  name: string;
  width: number;
  height: number;
  image: string;
  map: string[][];
  positivesList: number[][] = [];
  negativesList: number[][] = [];

  private populateLists() {
    this.map.forEach((line, lineIndex) => {
      for (let pixel = 0; pixel < line.length; pixel++) {
        if (line[pixel] === "o") {
          this.positivesList.push([lineIndex, pixel]);
        } else {
          this.negativesList.push([lineIndex, pixel]);
        }
      }
    });
  }

  constructor(name: string, image: string,) {
    this.image = image.trim();
    this.name = name;
    this.map = this.processImage();
    this.width = this.map[0].length;
    this.height = this.map.length;
    this.populateLists();
  }

  processImage() {
    const linesArray = this.image.split("\n");
    return linesArray.map((line) => line.split(""));
  }



}
