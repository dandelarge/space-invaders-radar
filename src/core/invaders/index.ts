export * from './Invader';
export * from './Spider';
export * from './Squid';
