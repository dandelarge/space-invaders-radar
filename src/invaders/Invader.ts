export class Invader {
  width: number;
  height: number;
  image: string;
  map: string[][];
  positivesList: number[][] = [];
  negativesList: number[][] = [];

  private populateLists() {
    this.map.forEach((line, lineIndex) => {
      for (let pixel = 0; pixel < line.length; pixel++) {
        if (line[pixel] === "o") {
          this.positivesList.push([lineIndex, pixel]);
        } else {
          this.negativesList.push([lineIndex, pixel]);
        }
      }
    });
  }

  constructor(image: string) {
    this.image = image.trim();
    this.map = this.processImage();
    this.width = this.map[0].length;
    this.height = this.map.length;
    this.populateLists();
  }

  processImage() {
    const linesArray = this.image.split("\n");
    return linesArray.map((line) => line.split(""));
  }

  static calculateMatchPercentage(
    sample: string[][],
    positives: number[][],
    negatives: number[][]
  ): {
    matchRatio: number;
    positives: number[][];
    negatives: number[][];
    sample: string[][];
  } {
    let matches = 0;

    const positivesMatches: number[][] = [];
    const negativesMatches: number[][] = [];

    positives.forEach((px) => {
      if (sample[px[0]][px[1]] === "o") {
        matches++;
        positivesMatches.push(px);
      }
    });

    negatives.forEach((px) => {
      if (sample[px[0]][px[1]] === "-") {
        matches++;
        negativesMatches.push(px);
      }
    });

    const matchRatio = matches / (sample.length * sample[0].length);

    return { matchRatio, positives: positivesMatches, negatives: negativesMatches, sample };
  }

}
