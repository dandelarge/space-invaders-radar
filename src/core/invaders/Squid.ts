import { Invader } from "./Invader";

const squidImage =
`---oo---
--oooo--
-oooooo-
oo-oo-oo
oooooooo
--o--o--
-o-oo-o-
o-o--o-o`;

export class Squid extends Invader {
    constructor() {
        super('squid', squidImage);
    }
}