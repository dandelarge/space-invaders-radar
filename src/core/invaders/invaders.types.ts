import { Invader } from "./Invader";

export interface RadarResults {
    matchRatio: number;
    positives: number[][];
    negatives: number[][];
    sample: string[][];
    image: string;
    invader: Invader;
    position: number[];
}