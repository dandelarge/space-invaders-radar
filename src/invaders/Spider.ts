import { Invader } from "./Invader";

const spiderImage =
`--o-----o--
---o---o---
--ooooooo--
-oo-ooo-oo-
ooooooooooo
o-ooooooo-o
o-o-----o-o
---oo-oo---`;

export class Spider extends Invader {
    constructor() {
        super(spiderImage);
    }
}