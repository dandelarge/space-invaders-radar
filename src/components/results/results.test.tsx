import React from 'react';
import { shallow } from 'enzyme';
import Results from './results';
import { RadarResults, RadarSample, RadarState } from '../../core/radar.types';
import { Invader } from '../../core/invaders/Invader';

const mockInvader = new Invader('mockInvader', '');

const mockSample: RadarSample = {
    position: [0,0],
    sample: [['o','-','0'],['o','-','0'],['o','-','0']]
};

const mockResult: RadarResults = {
    positives: [[0,0]],
    negatives: [[0,1]],
    sample: mockSample,
    matchRatio: 1,
    image: '',
    invader: mockInvader
};

const mockState: RadarState = {
    input:'',
    samples: [mockSample],
    results: [mockResult, mockResult],
    definition: 0,
    invaders:[]
};

test('renders the component', () => {
  const component = shallow(<Results state={mockState} />);
  expect(component).toMatchSnapshot();
});

test('should update the index on next button', () => {
    const component = shallow(<Results state={mockState} />);
});