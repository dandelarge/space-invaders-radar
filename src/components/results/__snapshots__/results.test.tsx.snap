// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<Fragment>
  <div>
    <div
      className="radarScreen"
    >
      <p
        className="textarea larger-text"
      >
        invaders found: 
        1
      </p>
    </div>
    <h4>
      current invader 
    </h4>
    <div>
      <button
        onClick={[Function]}
      >
        previous
      </button>
      <span>
        1
         of 
        1
      </span>
      <button
        onClick={[Function]}
      >
        next
      </button>
    </div>
  </div>
  <div
    className="radarScreen"
  >
    <p>
      Invader image from the radar in positition:
       
      0
       ,
       
      0
    </p>
    <pre
      className="textarea"
    />
  </div>
  <div
    className="radarScreen"
  >
    <p>
      Invader image taking in account only possitive matches
    </p>
    <pre
      className="textarea"
    >
      

    </pre>
  </div>
  <div
    className="radarScreen"
  >
    <p>
      Invader image removing only the noise
    </p>
    <pre
      className="textarea"
    >
      

    </pre>
  </div>
</Fragment>
`;
